circuit/src/merkle_tree/native.rs

```diff
@@ -0,0 +1,158 @@
+use std::marker::PhantomData;
+
+use halo2wrong::halo2::arithmetic::FieldExt;
+
+use crate::{params::RoundParams, poseidon::native::sponge::PoseidonSponge};
+
+const WIDTH: usize = 5;
+
+#[derive(Clone, Debug)]
+struct MerkleTree<F: FieldExt, P>
+where
+	P: RoundParams<F, WIDTH>,
+{
+	_params: PhantomData<P>,
+	root: Option<F>,
+	indexes: Vec<Option<F>>,
+	height: usize,
+	first_leaf_index: usize,
+}
+
+impl<F: FieldExt, P> MerkleTree<F, P>
+where
+	P: RoundParams<F, WIDTH>,
+{
+	fn new(height: usize) -> Self {
+		let mut first_leaf_index = 1;
+		let mut nodes = 1;
+
+		for i in 0..height {
+			first_leaf_index *= 2;
+			nodes = 2 * nodes;
+		}
+		first_leaf_index -= 1;
+		nodes *= 2;
+		let indexes = vec![None; nodes - 1];
+		MerkleTree { _params: PhantomData, root: None, indexes, height, first_leaf_index }
+	}
+
+	fn implement_leaves(self: &mut MerkleTree<F, P>, values: Vec<F>) -> &Self {
```

Comment:
Not sure the name `implement_leaves` is the best choice. Add leaves sounds more clear